## Funnel Analysis
I came across a few small datasets from an e-commerce company, I thought it would be fun to quickly perform a funnel analysis.  
I begin by loading and inspecting the datasets.

In [64]:
import pandas as pd

In [65]:
purchase = pd.read_csv('purchase.csv')
visits = pd.read_csv('visits.csv')
cart = pd.read_csv('cart.csv')
checkout = pd.read_csv('checkout.csv')

In [66]:
print(cart.shape)
cart.head(5)

(348, 2)


user_id            cart_time
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00
1  4397f73f-1da3-4ab3-91af-762792e25973  2017-05-27 01:35:00
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663  2017-03-04 10:38:00
3  b594862a-36c5-47d5-b818-6e9512b939b3  2017-09-27 08:22:00
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  2017-07-26 15:48:00

In [67]:
print(checkout.shape)
checkout.head(5)

(226, 2)


user_id        checkout_time
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d  2017-06-25 09:29:00
1  4ac186f0-9954-4fea-8a27-c081e428e34e  2017-04-07 20:11:00
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d  2017-07-13 11:38:00
3  89fe330a-8966-4756-8f7c-3bdbcd47279a  2017-04-20 16:15:00
4  3ccdaf69-2d30-40de-b083-51372881aedd  2017-01-08 20:52:00

In [68]:
print(purchase.shape)
purchase.head(5)

(252, 2)


user_id        purchase_time
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85  2017-05-11 04:25:00
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac  2017-09-05 08:45:00
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc  2017-11-20 20:49:00
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d  2017-01-22 15:18:00
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867  2017-06-11 11:32:00

In [69]:
print(visits.shape)
visits.head(5)

(2000, 2)


user_id           visit_time
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00

With the datasets loaded, I wanted to determine what percentage of shoppers visited the site but did not place anything in their shopping cart. I begin by performing a left merge on the *visits* and *cart* datasets.  
Of the 2000 visitors to the site, I have determined that 1652 (82.6%) of them did not place anything in their shopping cart.

In [70]:
visit_cart = pd.merge(visits, cart, how='left')
print("Number of observations: ", len(visit_cart))
visit_cart.head(5)

Number of observations:  2000


user_id           visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   

             cart_time  
0                  NaN  
1  2017-01-26 14:44:00  
2  2017-08-20 08:31:00  
3                  NaN  
4                  NaN

In [71]:
null = len(visit_cart[visit_cart.cart_time.isnull()])
no_cart = round((null / len(visit_cart)) * 100, 2)
print("Number of shoppers who did not place an item in their shopping cart: ", null)
print('Percentage of shoppers who did not place an item in their shopping cart: ', no_cart, '%')

Number of shoppers who did not place an item in their shopping cart:  1652
Percentage of shoppers who did not place an item in their shopping cart:  82.6 %


Next I set out to determine what percentage of shoppers placed items in their shopping cart but later abandoned it.  
I have determined that out of the 348 visitors who placed items in their shopping cart, 122 carts (35.06%) were abandoned.

In [72]:
cart_checkout = pd.merge(cart, checkout, how='left')
print("Number of observations: ", len(cart_checkout))
cart_checkout.head(5)

Number of observations:  348


user_id            cart_time  \
0  2be90e7c-9cca-44e0-bcc5-124b945ff168  2017-11-07 20:45:00   
1  4397f73f-1da3-4ab3-91af-762792e25973  2017-05-27 01:35:00   
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663  2017-03-04 10:38:00   
3  b594862a-36c5-47d5-b818-6e9512b939b3  2017-09-27 08:22:00   
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974  2017-07-26 15:48:00   

         checkout_time  
0  2017-11-07 21:14:00  
1                  NaN  
2  2017-03-04 11:04:00  
3  2017-09-27 08:26:00  
4                  NaN

In [73]:
null_cart_checkout = len(cart_checkout[cart_checkout.checkout_time.isnull()])
no_checkout = round((null_cart_checkout / len(cart_checkout) * 100), 2)
print("Number of abandoned carts: ", null_cart_checkout)
print("Percentage of shoppers who abandoned their cart: ", no_checkout, "%")

Number of abandoned carts:  122
Percentage of shoppers who abandoned their cart:  35.06 %


In this section, I wanted to determine what percentage of visitors placed items in their cart, proceeded to checkout, but did not end up making the purchase. 
I have determined that of the 334 visitors who proceeded to check out, 82 (24.55%) of them did not complete their purchase.

In [74]:
all_data = visit_cart.merge(cart_checkout, how='left').merge(purchase, how='left')
all_data.head(5)

user_id           visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188  2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2  2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1  2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab  2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96  2017-02-27 11:25:00   

             cart_time        checkout_time        purchase_time  
0                  NaN                  NaN                  NaN  
1  2017-01-26 14:44:00  2017-01-26 14:54:00  2017-01-26 15:08:00  
2  2017-08-20 08:31:00                  NaN                  NaN  
3                  NaN                  NaN                  NaN  
4                  NaN                  NaN                  NaN

In [84]:
checkout_no_purchase = len(all_data[(~all_data.checkout_time.isnull()) & (all_data.purchase_time.isnull())])
checkout_no_purchase_percent = round((checkout_no_purchase / len(all_data[~all_data.checkout_time.isnull()]) * 100), 2)
print("Total number of shoppers who proceeded to checkout: ", len(all_data[~all_data.checkout_time.isnull()]))
print("Number of shoppers who proceeded to checkout but did not purchase: ", checkout_no_purchase)
print("Percentage of shoppers who proceeded to checkout but did not purchase: ", checkout_no_purchase_percent, "%")

Total number of shoppers who proceeded to checkout:  334
Number of shoppers who proceeded to checkout but did not purchase:  82
Percentage of shoppers who proceeded to checkout but did not purchase:  24.55 %
